### PR TITLE
#2022 Fine Grained HTTP Compliance Modes

### LINKED ISSUE #2022: Illegal character 0x20 in state=HEADER for buffer
Hi,
we are using the the jetty HTTP client in our server to request external resources. But we get some strange errors on some pages since they do not behave compatible to the latest HTTP specs. 

For instance if you request [https://us.braun.com/en-us](https://us.braun.com/en-us) we get following logs:
`Illegal character 0x20 in state=HEADER for buffer DirectByteBuffer@7b362163[p=590,l=13192,c=16384,r=12602]={HTTP/1.1 200 OK\r\n...-Allow-Headers <<<: Origin, X-Reque...4E\xFd\xF0\xF84z \xF4+|\xF4\x95=\x1b>>>\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00...\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00}`

The Response headers looks as the following:
```
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store
Pragma: no-cache
Content-Type: text/html; charset=utf-8
Expires: -1
Vary: User-Agent, Accept-Encoding
Server: Microsoft-IIS/8.5
Set-Cookie: braun_us#lang=en-US; path=/
Set-Cookie: ASP.NET_SessionId=pimb44eu1ts4sd1raqs2sl4t; path=/; HttpOnly
Strict-Transport-Security: max-age=300
X-Frame-Options: SAMEORIGIN
Set-Cookie: __CSRFCOOKIE=m71fgPdokWdedn4Lt3TyntwmLA5QtPDEiFIXZRVByhalmgOTVboj5BP; path=/; HttpOnly
Access-Control-Allow-Headers : Origin, X-Requested-With, Content-Type, Accept
Access-Control-Allow-Origin: *
X-Frame-Options: SAMEORIGIN
X-UA-Compatible: IE=edge
X-Server: WPSC38
X-Content-Type-Options: nosniff
Date: Tue, 05 Dec 2017 20:51:25 GMT
Set-Cookie: BIGipServer~MAN3-PNG-RD1~us.braun.com-SC-ares.com=rd1o00000000000000000000ffff0a8bc13co8651; path=/
Content-Encoding: gzip
Transfer-Encoding: chunked
```

Affected Jetty Version:
9.3 (Tested with 9.3.14.v20161028)
9.4 (Tested with 9.4.8.v20171121)

Note that the issue does not happen always on that webpage.

The error seems to be in the [HttpParser](https://github.com/eclipse/jetty.project/blob/d4061fcfeb5a959e67547404a1fdfa5a203680db/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java#L1290) since a space is not expected in the HTTP header. However, there exist Servers which misbehave here and I expect that the Jetty Client should communicate with those Servers.

Do you see a way how Jetty can handle this properly? I see that there already exists many compliance checks in the Parser. Can this be added too?

### PR DESCRIPTION
**Pull request by gregw**
Fixes #2022 
Taking over from #2060, so we can work on better compliance modes

### CODE REVIEW

**Comment by gregw** [comment]
As usual, the IP validation check is unable to deal with mixed PRs like this! the contributions from @fbuecklers are correctly signed-off and I'm a committer, so not idea why the merge breaks IP validation.  So I will manually check and take responsibility before merging...  (must remember not to squash!!!)

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java [hunk 1]
```diff
@@ -31,29 +36,122 @@
...
-            HttpRFC.RFC7230_3_2_4_WS_AFTER_FIELD_NAME,
-            HttpRFC.RFC7230_3_2_4_NO_FOLDING,
-            HttpRFC.RFC7230_A2_NO_HTTP_9))), 
-    RFC7230(EnumSet.allOf(HttpRFC.class)), 
-    ;
+    // TODO in Jetty-10 convert this enum to a class so that extra custom modes can be defined dynamically
+    LEGACY(sectionsBySpec("LEGACY")), 
+    RFC2616(sectionsBySpec("RFC2616")), 
+    RFC7230(sectionsBySpec("RFC7230,-RFC7230_3_1_1_METHOD_CASE_SENSITIVE")), // TODO fix in Jetty-10
+    CUSTOM0(sectionsByProperty("CUSTOM0")),
```
Comment (joakime) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: Why are these here?
Why can't someone just create a HttpCompliance object instead? (change from enum to class with static definitions)

Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: Because currently HttpCompliance is an enum.  To make it dynamic so that anyone can add new modes, would require it to be a class.   This can be done in Jetty-10, but it is too much of a breaking change for 9.4.x as it would break compiled code.

I do have a TODO in there about how it should be converted to a class in jetty-10

**Comment by gregw** [comment]
@fbuecklers Can you have a look at my changes and see of the custom compliance modes make sense for you.... oh I'll add some javadoc on how to use them with a commit in 20 minutes or so....

**Comment by gregw** [comment]
I've split the modes into `RFC2616`, `RFC2616_STRICT`, `RFC7230` & `RFC7230_STRICT` to capture the difference from how the modes should be vs how we've historically supported them.

Alternately we could define: `RFC2616`, `RFC2616_LEGACY`, `RFC7230` and `RFC7230_LEGACY` to set the main modes to the correct sets, but provide our legacy support settings.    

@joakime do you think that would be OK in a 9.4.x release? It would change default behaviour but in minor ways while providing easy backwards compatibility.    Actually, I think we should define them that way and then just decide if the default in 9.4.x should be `RFC7230` or `RFC7230_LEGACY`?

**Comment by gregw** [comment]
I've followed my own advice and changed the modes to `RFC7230` and `RFC7230_LEGACY` with the default being the later.   This keeps the behaviour of the standard distribution, but is a small change in behaviour for embedded usages.   As these are changes to comply with the RFC, I think this is an acceptable compromise.

**Comment by fbuecklers** [comment]
@gregw, In General, it looks good for me and it works. I have configured the client with the following code:
```
HttpCompliance compliance = HttpCompliance.CUSTOM0;
compliance.sections().clear();
compliance.sections().add(HttpComplianceSection.RFC7230_3_2_CASE_INSENSITIVE_FIELD_NAME);
compliance.sections().add(HttpComplianceSection.USE_CASE_INSENSITIVE_FIELD_VALUE_CACHE);

this.httpClient = createClient(pool);
this.httpClient.setHttpCompliance(compliance);
```

It seems to be a little bit ugly that I have to clear the `HttpCompliance.CUSTOM0` list beforehand. In general, I think it would be useful to have a compliance mode that is as weak as possible. This will ensure that the jetty client behaves similarly to real browsers. They are also loading and displaying web pages which has many HTTP compliance issues by just ignoring those errors.

**Comment by gregw** [comment]
@fbuecklers I think the best starting point for the CUSTOM enums is complete compliance with the latest RFC.  In Jetty-10, I shall modify the the enum to be a class and provide constructors that will allow full or empty section lists as a starting point.

**Comment by gregw** [comment]
@janbartel @sbordet @joakime @jmcc0nn3ll @olamy @WalkerWatch  I've added all of you as reviewers as I'm super cautious about merging this change.  Once done it (or something based on it) will be with us forever.  Thus I'd like to get everybody's consensus before merging.

**Comment by joakime** [comment]
For those of you reviewing this PR, here's your reading list ...

* Issue #2022 - Illegal character 0x20 in state=HEADER for buffer (the issue that triggered this PR)
* PR #2060 - Illegal character 0x20 in state=HEADER (old, unmerged, PR that this one is based off of)

**Comment by gregw** [comment]
@janbartel @sbordet @joakime @jmcc0nn3ll @olamy @WalkerWatch Your thoughts????

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java [hunk 2]
```diff
@@ -18,15 +18,171 @@
...
+    RFC2616(sectionsBySpec("RFC2616")), 
+    
+    /** Jetty's current RFC7230 support, which incorrectly excludes  {@link HttpComplianceSection#RFC7230_3_1_1_METHOD_CASE_SENSITIVE} */
+    RFC7230_LEGACY(sectionsBySpec("RFC7230,-RFC7230_3_1_1_METHOD_CASE_SENSITIVE")),
+
+    /** The RFC7230 support mode */
+    RFC7230(sectionsBySpec("RFC7230")),
+    
+    /** Custom compliance mode that can be defined with System property <code>org.eclipse.jetty.http.HttpCompliance.CUSTOM0</code> */
+    CUSTOM0(sectionsByProperty("CUSTOM0")),
```
Comment (joakime) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: The CUSTOM0 enums/modes are awkward.
We shouldn't encourage this form of usage.

**Comment by jmcc0nn3ll** [comment]
I like the general direction, the enum convention seems forward compatible and given feedback over the last few years we really need a solid convention for this area.  I am having a hard time wrapping my head around impact to people using existing compliance checking though.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpComplianceSection.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+//
+
+package org.eclipse.jetty.http;
+
+/**
+ */
+public enum HttpComplianceSection 
+{     
+    USE_CASE_INSENSITIVE_FIELD_VALUE_CACHE("","Use case insensitive field value cache"),
+    RFC7230_3_1_1_METHOD_CASE_SENSITIVE("https://tools.ietf.org/html/rfc7230#section-3.1.1","Method is case-sensitive"),
```
Comment (joakime) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpComplianceSection.java]: This name is awkward.
What happens when RFC7230 is replaced?
I propose that the enum doesn't contain the RFC number and the section number, let the url have that.
Change it to the technology name of the spec.

Eg:

Current Name | New Name
-------------- | ------------
RFC7230_3_1_1_METHOD_CASE_SENSITIVE | HTTP_11_METHOD_CASE_SENSITIVE
RFC7230_3_2_FIELD_COLON | HTTP_11_FIELD_COLON
RFC7230_3_2_CASE_INSENSITIVE_FIELD_NAME | HTTP_11_CASE_INSENSITIVE_FIELD_NAME
RFC7230_3_2_4_NO_WS_AFTER_FIELD_NAME | HTTP_11_NO_WHITESPACE_AFTER_FIELD_NAME
RFC7230_3_2_4_NO_FOLDING | HTTP_11_NO_LINE_FOLDING
RFC7230_A2_NO_HTTP_9 | HTTP_09_NOT_SUPPORTED

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java [hunk 4]
```diff
@@ -714,7 +780,7 @@ public void testCaseInsensitiveHeader() throws Exception
                         "cOnNeCtIoN: ClOsE\r\n" +
                         "\r\n");
         HttpParser.RequestHandler handler = new Handler();
-        HttpParser parser = new HttpParser(handler, -1, HttpCompliance.RFC7230);
+        HttpParser parser = new HttpParser(handler, -1, HttpCompliance.RFC7230_LEGACY);
```
Comment (joakime) [inline jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java]: We need to keep the older RFC7230 tests too, to ensure we didn't break backward compliance with older 9.4.x releases.

Comment (joakime) [reply jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java]: Perhaps these tests should be parameterized?

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java (hunk 2 above)
Comment (joakime) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: Since the implementation in HttpParser seems to only use the HttpComplianceSection rules now, why can't this HttpCompliance just be a predefined set of rules to start from, but isn't the active set that is in use?

Proposed usage:
``` java
HttpComplianceConfig complianceConfig = new HttpComplianceConfig(HttpCompliance.RFC7230);
complianceConfig.removeSection(HttpComplianceSection.RFC7230_3_2_4_NO_WS_AFTER_FIELD_NAME);
HttpParser parser = new HttpParser(handler, complianceConfig);
```

Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: The CUSTOMn approach is a kludge for jetty-9.x, where we need to keep using an enum for backward compatibility.  For Jetty-10, we can use a class instead and deprecate the custom static fields.

Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: While it would be nice to be able to pass in to the HttpParser some kind of programatic collection of HttpComplianceSection, that would break backwards compatibility.  We currently set the compliance mode with an enum and that is in several APIs and XMLs.  This approach keeps that enum, but just makes it a holder of collections of HttpComplianceSections.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpComplianceSection.java (hunk 3 above)
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpComplianceSection.java]: Good idea!  Actually do we even need HTTP_11_ prefix?

The other thing I've tossed around is should they be negations or enablers?  ie should it just be LINE_FOLDING and we don't add that for RFC7230 compliance.  That kind of feels better, but then the URL reference makes no sense as it points to a section that says it is not supported.

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java (hunk 4 above)
Comment (gregw) [reply jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java]: Perhaps, but at this stage I'd like to keep the tests pretty much as is, so that we don't inadvertently  change behaviour (but perhaps a few little advertent changes :)

**Comment by gregw** [comment]
@jmcc0nn3ll The impact on users of the current compliance should be very minimal.  They currently set and enum and they can continue to set an enum for compliance.  We have "fixed" the compliance of `RFC7230`, but we provide the `RFC7230_LEGACY` enum if they really want our old bugs to persist

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpComplianceSection.java (hunk 3 above)
Comment (WalkerWatch) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpComplianceSection.java]: I think it is worth including the technology name, just in case there is ever a change to specific functionality in the future where we have diverging but similar behaviors between, say, HTTP 1.1 and HTTP 2 that we need to differentiate from. I can't see this being common but feels like it might help future proof it some and possibly be easier to identify at a casual glance? Maybe I am being too pedantic.

Comment (jmcc0nn3ll) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpComplianceSection.java]: @WalkerWatch we should have a compliance page of its own that lists out all the options and what they represent

Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpComplianceSection.java]: @WalkerWatch problem with technology names is that it is hard to allocated. Some sections apply to just HTTP/1 (Eg white space after field name), some can apply to HTTP/1 and HTTP/2 (eg case insensitivity).    We also don't know how future specs will go, so we don't know if they will change.

Comment (joakime) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpComplianceSection.java]: Our use of these is currently technology specific.
I can see someone wanting HTTP/1.1 to be more strict then HTTP/1.0 (for example), so having two different "levers" to pull would be a good thing.

**Comment by gregw** [comment]
@janbartel @sbordet @joakime @jmcc0nn3ll @olamy @WalkerWatch  Can somebody please either give me a review OK so I can merge or tell me what needs to be fixed!  Let's not leave this hanging for yet another week!!!!!!

File: jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java [hunk 5]
```diff
@@ -38,13 +39,15 @@
...
 {
-    private final HttpParser parser = new HttpParser(this);
+    private final HttpParser parser;
     private ByteBuffer buffer;
     private boolean shutdown;
 
     public HttpReceiverOverHTTP(HttpChannelOverHTTP channel)
     {
         super(channel);
+        // TODO: Seems to be not the optimal way to get the compliance mode
```
Comment (janbartel) [inline jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java]: Is this really a TODO, or something that we should resolve in this PR?

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java [hunk 6]
```diff
@@ -18,15 +18,171 @@
...
+ * list of sections.  The list should start with one of the following strings:<dl>
+ * <dt>0</dt><dd>No {@link HttpComplianceSection}s</dd>
+ * <dt>*</dt><dd>All {@link HttpComplianceSection}s</dd>
+ * <dt>RFC2616</dt><dd>The set of {@link HttpComplianceSection}s application to https://tools.ietf.org/html/rfc2616,
+ * but not https://tools.ietf.org/html/rfc7230</dd>
+ * <dt>RFC7230</dt><dd>The set of {@link HttpComplianceSection}s application to https://tools.ietf.org/html/rfc7230</dd>
  * </dl>
+ * The remainder of the list can contain then names of {@link HttpComplianceSection}s to include them in the mode, or prefixed
+ * with a '-' to exclude thm from the mode.    Note that Jetty's modes may have some historic minor differences from the strict
+ * RFC compliance, for example the <code>RFC2616</code> HttpCompliance is defines as
```
Comment (janbartel) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: should be "is defined as" , not "is defines as"

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java [hunk 7]
```diff
@@ -18,15 +18,171 @@
...
+ * with a '-' to exclude thm from the mode.    Note that Jetty's modes may have some historic minor differences from the strict
+ * RFC compliance, for example the <code>RFC2616</code> HttpCompliance is defines as 
+ * <code>RFC2616,-RFC7230_3_2_FIELD_COLON,-RFC7230_3_1_1_METHOD_CASE_SENSITIVE</code>.
+ * <p>
+ * Note also that the {@link EnumSet} return by {@link HttpCompliance#sections()} is mutable, so that modes may 
+ * be altered in code and will affect all usages of the mode.
  */
-public enum HttpCompliance { LEGACY, RFC2616, RFC7230 }
\ No newline at end of file
+public enum HttpCompliance // TODO in Jetty-10 convert this enum to a class so that extra custom modes can be defined dynamically
```
Comment (janbartel) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: Instead of a todo, how about just making an issue?

File: jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java [hunk 8]
```diff
@@ -336,13 +336,6 @@ protected ServerSocketChannel openAcceptChannel() throws IOException
 
         return serverChannel;
     }
-    
-    @Override
-    public Future<Void> shutdown()
-    {
-        // shutdown all the connections
-        return super.shutdown();
-    }
```
Comment (janbartel) [inline jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java]: How come this got removed by this PR?

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java [hunk 9]
```diff
@@ -112,15 +120,15 @@ public void testLineParse1_RFC2616() throws Exception
         ByteBuffer buffer = BufferUtil.toBuffer("GET /999\r\n");
 
         HttpParser.RequestHandler handler = new Handler();
-        HttpParser parser = new HttpParser(handler, HttpCompliance.RFC2616);
+        HttpParser parser = new HttpParser(handler, HttpCompliance.RFC2616_LEGACY);
```
Comment (janbartel) [inline jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java]: This is probably not the best place for this comment, but  ....

So, IIUC, folks who were using HttpCompliance.RFC2616 now have to use HttpCompliance.RFC2616_LEGACY to get the same behaviour???  If so, I think I prefer the other alternative of RFC2616_STRICT for a strict adherence to the spec, so no user code changes are required.

Comment (joakime) [reply jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java]: @janbartel has a point here with backward compatibility with existing users.

+1 for _STRICT

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java (hunk 7 above)
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java]: I'll do both.  I'll create the issue IFF this PR ever gets accepted :)

File: jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java (hunk 8 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java]: It's a strange change, not related to the scope of the issue, but it's ultimately harmless.

It just calls `super.shutdown();` 

[AbstractConnector.shutdown()](https://github.com/eclipse/jetty.project/blob/f046bdc2c61584462949ea9aee10b32a43bb74d6/jetty-server/src/main/java/org/eclipse/jetty/server/AbstractNetworkConnector.java#L102-L107) is where the work is anyway.

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java (hunk 9 above)
Comment (gregw) [reply jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java]: I disagree.  It is a bug that our RFC2616 compliance mode was not actually RFC2616 compliant.  Users that select that mode want compliance with the RFC and not with a legacy interpretation of it by jetty.  As we find/fix bugs with our RFC compliance, we need to be able to update the RFC compliance modes.  Users that wish to be compatible with bugs can either edit the compliance modes or they can use a legacy mode.

File: jetty-http/src/test/resources/jetty-logging.properties [hunk 10]
```diff
@@ -1,3 +1,3 @@
 org.eclipse.jetty.util.log.class=org.eclipse.jetty.util.log.StdErrLog
 #org.eclipse.jetty.LEVEL=DEBUG
-#org.eclipse.jetty.server.LEVEL=DEBUG
+org.eclipse.jetty.server.LEVEL=DEBUG
```
Comment (joakime) [inline jetty-http/src/test/resources/jetty-logging.properties]: Noisy logging during testing (btw)

File: jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java (hunk 8 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java]: because it is a noop.  I can put it back if you really want.... (yes I'm breaking my own rules for putting in unrelated changes).

File: jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java [hunk 11]
```diff
@@ -972,6 +975,25 @@ public void setDispatchIO(boolean dispatchIO)
...
+    {
+        return httpCompliance;
+    }
+
+    /**
+     * Sets the http compliance mode for parsing http responses.
+     * This affect how weak the {@link HttpParser} parses http responses and which http protocol level is supported
+     * @param httpCompliance The compliance level which is used to actually parse http responses
+     */
+    public void setHttpCompliance(HttpCompliance httpCompliance)
```
Comment (joakime) [inline jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java]: This is a new API for HttpClient, can't we do this correctly here?
A `Collection<HttpComplianceSection>` instead?

Comment (gregw) [reply jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java]: `HttpCompliance` is-a `Collection<HttpComplianceSection>`, it is just a really clunky non-dynamic one.    The future is for HttpCompliance to be a class, so this API is future correct.

Comment (joakime) [reply jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java]: I don't follow.

I thought the `HttpCompliance` enum was a convenient location for pre-defined (immutable?)  `Collection<HttpComplianceSection>` baselines.  Something you can start with then customize.

Comment (gregw) [reply jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java]: This PR changes `HttpCompliance` from being a single valued enum, to being an enum that contains a Collection<HttpComplianceSection>.   So when you pass a HttpCompliance you get two things: a name (instance) for the mode you are running in AND a `Collection<HttpComplianceSection>`.

This is all good.... except because it is an enum we can only have a precompiled fixed number of such modes.  So we have had to create the custom modes and make them mutable.

In future, we will not be passing around raw `Collection<HttpComplianceSection>` as we will lose the name of the mode.   We will still pass around `HttpCompliance`s instances, it is just that they will be classes rather than enums, so users can define their own.   There will still be `HttpCompliance.RFC7230` constants, but they will be static instances rather than enums - thus the code will be compile compatible, but not runtime compatible.

**Comment by gregw** [comment]
@joakime @janbartel  I believe I have fixed or addressed your concerns. Can you please dismiss your X or   respond so I can get this merged.
